bundles/org.openhab.binding.lgwebos/src/main/java/org/openhab/binding/lgwebos/internal/discovery/LGWebOSUpnpShutdownDetector.java

```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.lgwebos.internal.discovery;
+
+import static org.openhab.binding.lgwebos.internal.LGWebOSBindingConstants.*;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.library.types.OnOffType;
+import org.eclipse.smarthome.core.thing.ThingRegistry;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.jupnp.UpnpService;
+import org.jupnp.model.meta.LocalDevice;
+import org.jupnp.model.meta.RemoteDevice;
+import org.jupnp.model.meta.RemoteService;
+import org.jupnp.registry.Registry;
+import org.jupnp.registry.RegistryListener;
+import org.openhab.binding.lgwebos.internal.handler.LGWebOSHandler;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class detects TV shutdown much before TV closes the websocket by listening to the upnp registry.
+ *
+ * Upnp devices send a good-bye broadcast on normal shutdown. This is true also for webos devices.
+ * This seems to be the only way to detect instantly that the TV was turned off by remote control and much before the
+ * device actually closes the websocket connection.
+ *
+ * However, not all users do use Upnp, so this use case is an optional optimization.
+ *
+ * The only Upnp device in webos that sends an event on shut down is of type
+ * urn:schemas-upnp-org:device:MediaRenderer:1, with services types:
+ * urn:schemas-upnp-org:service:AVTransport:1
+ * urn:schemas-upnp-org:service:ConnectionManager:1
+ * urn:schemas-upnp-org:service:RenderingControl:1
+ *
+ * 20 to 30 seconds later the websocket connection will close.
+ * Arround this point in time device type: urn:schemas-upnp-org:device:Basic:1 de-registers with service type:
+ * urn:lge-com:service:webos-second-screen:1 - the same, which is used for thing discovery.
+ * Thus, thingRemoved of DiscoveryListener implementation in LGWebOSHandler will only be called at this point.
+ *
+ *
+ * @author Sebastian Prehn - Initial contribution
+ *
+ */
+@NonNullByDefault
+@Component(immediate = true, configurationPid = "discovery.lgwebos.shutdown")
+public class LGWebOSUpnpShutdownDetector implements RegistryListener {
```

Comment:
Rather than reimplementing something from the core framework (using directly JUPnP stuff) and creating a dependency between your binding and the thing registry (which is normally unexpected), why not updating your discovery service to consider urn:schemas-upnp-org:device:MediaRenderer:1 rather than urn:schemas-upnp-org:device:Basic:1 ?

Reply:
I noticed as well, but how can I reliably detect LG WebOS devices only? 

The connectsdk implementation from LG did search for urn:schemas-upnp-org:device:Basic:1 and this worked reliably in the past. 

If we react to urn:schemas-upnp-org:device:MediaRenderer:1 how can we make sure we filter for webos. I don't have any documentation from LG on what would be sent in meta data every time.